experimental/Schemes/Auxiliary.jl

```diff
@@ -19,46 +19,111 @@ function pullback(f::CoveredSchemeMorphism)
       pbf = pullback(f_U)
       ID[U] = ideal(OO(U), pbf.(gens(II(V))))
     end
-    return IdealSheaf(X, ID, check=false) # TODO: Set to false eventually.
+    return IdealSheaf(X, ID, check=false)
   end
 
-  function pullback_func(C::CartierDivisor)
+  function pullback_func(C::EffectiveCartierDivisor)
     phi = covering_morphism(f)
     triv_dict = IdDict{AbsSpec, RingElem}()
-    OOX = OO(X)
-    # We need to do the following:
-    # The divisor D := f^* C needs to be trivialized on some refinement. 
-    # The one we're using consists of the preimages of the patches of 
-    # the `trivializing_covering` of `C`. These must be intersected 
-    # with each and every affine chart of X first. 
-    for U in patches(domain(phi))
-      V = codomain(phi[U])
-      for W in patches(trivializing_covering(C))
-        # In case V and W do not happen to lay in the same affine 
-        # chart, the preimage of W in U will be empty. So we can 
-        # skip the computation in that case.
-        success, par = _have_common_ancestor(V, W)
-        if success
-          # Reconstruct W as a PrincipalOpenSubset of the affine chart directly
-          # (Note that we might have gone through some calls to `simplify(...)`, 
-          # so this really is a non-trivial step)
-          inc_W_flat = _flatten_open_subscheme(W, par)
-          # To cheaply construct the preimage of W in U, just pull back the 
-          # complement equation.
-          h = complement_equation(codomain(inc_W_flat))
-          UW = PrincipalOpenSubset(U, pullback(phi[U])(OOX(par, V)(h)))
-          # Finally, we can add the trivialization on U ∩ f⁻¹(W) to our list.
-          triv_dict[UW] = pullback(phi[U])(first(C(V)))
-        end
-      end
+    psi = restrict(f, trivializing_covering(C))
+    triv_cov = domain(psi)
+    for U in patches(triv_cov)
+      V = codomain(psi[U])
+      pbgens = pullback(psi[U]).(C(V))
+
+      # Do the sanity checks
+      length(pbgens) == 1 || error("cartier divisor is not principal on this patch")
+      g = first(pbgens)
+      # See the Stacks project on cartier divisors
+      is_zero_divisor(g) && error("pullback of the local equation is a zero divisor; pullback of the cartier divisor is not defined")
+      triv_dict[U] = g
     end
-    return CartierDivisor(X, triv_dict) # We have to leave the checks back on, because local 
-                                        # equations might pull back to zero or zero divisors.
+    
+    return EffectiveCartierDivisor(X, triv_dict, trivializing_covering=triv_cov, check=false)
+  end
+
+  function pullback_func(CC::Covering)
+    psi = restrict(f, CC)
+    return domain(psi)
   end
 
   return pullback_func
 end
 
+@Markdown.doc """
+    restrict(f::CoveredSchemeMorphism, DD::Covering)
```

Comment:
I would have called it refine. 
But since it is in experimental, I am good to go.

Reply:
The `_covering_morphism` does not really add anything new since it is the first argument anyways.
I think we lack the mathematical terminology here. So maybe we are free to use our own terminology. But we should be  consistent. 
General situation: Some $X$ is given on a covering $C$ and we have a refinement $D \to C$. 
We could denote that $X|D$ and say it is the restriction. Because in a sense, its patches are the restrictions. 
The most common example would probably be a section of a sheaf (including a hom sheaf).

Reply:
So unless we come up with better terminology I am okay with restriction.